polar-core/src/sources.rs

```diff
@@ -0,0 +1,66 @@
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Hash)]
+pub enum SourceInfo {
+    // From the parser
+    Parser {
+        /// Index into the source map stored in the knowledge base
+        src_id: u64,
+
+        /// Location of the term within the source map
+        left: usize,
+        right: usize,
+    },
+
+    /// Created as a temporary variable
+    TemporaryVariable,
+
+    /// From an FFI call
+    Ffi,
+
+    /// Created for a test
+    Test,
+}
+
+impl SourceInfo {
+    pub fn ffi() -> Self {
+        Self::Ffi
+    }
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct Source {
+    pub filename: Option<String>,
+    pub src: String,
+}
+
+pub struct Sources {
+    // Pair of maps to go from Term ID -> Source ID -> Source.
```

Comment:
comment needs to be updated, and is this struct still necessary?

Reply:
Comment updated. The struct itself is admittedly simple, but I think offers a worthwhile abstraction.

Reply:
just wary of abstraction leading to unnecessary layers of duplication, like the KB having a `sources` property that is a `Sources` that has a `sources` property 😄